Add screenshot teardown fixture - approach 1

danielhollas opened the pull request:
We want to take a screenshot at the end of each selenium test, even if (and especially) when the test fails.
So essentially we need to take the screenshot as part of the test teardown. Here's the relevant section of the pytest docs

https://docs.pytest.org/en/6.2.x/fixture.html#teardown-cleanup-aka-fixture-finalization

Our case is complicated by the fact that we somehow need to pass the screenshot name to the teardown method.
Here's the relevant section in docs and two SO questions were helpful:

https://docs.pytest.org/en/6.2.x/fixture.html#fixtures-can-introspect-the-requesting-test-context
https://stackoverflow.com/questions/58549670/pass-a-parameter-to-a-pytest-fixture-for-teardown

https://stackoverflow.com/questions/38488571/attributeerror-when-using-request-function-in-pytest-yield-fixture

In this PR, we take the approach of having a separate yield fixture. We pass the screenshot filename as a parameter to a dictionary object that is yielded by the fixture. Still feels a bit hacky, but it is quite clean and compact.

I'll also submit an PR with an alternative solution, where we extend the existing `selenium_driver` fixture.

Closes #419

unkcpz reviewed (commented):
thanks a lot! @danielhollas 
This is very interesting, and I prefer this approach as you said having a dedicated fixture is more clear and easy to maintain in the future. 

But is that possible to make this fixture call as a function by `screenshot("process-list.png")`?

unkcpz commented on tests/test_notebooks.py:
```diff
@@ -10,49 +10,49 @@ def test_notebook_service_available(notebook_service):
-def test_process(selenium_driver, screenshot_dir):
+def test_process(selenium_driver, screenshot):
     driver = selenium_driver("notebooks/process.ipynb")
+    screenshot["name"] = "process.png"
```
Is the fixture must be called at the very beginning or it can be moved to the last which makes more sense?

yakutovicha replied:
I think running this at the beginning is better. This line sets the screenshot filename and puts on hold the rest of the code that follows `yield`.

The code after `yield` will be executed once all the tests are completed. More information can be found following the [link](https://docs.pytest.org/en/6.2.x/fixture.html#yield-fixtures-recommended) shared by @danielhollas.

danielhollas commented:
> But is that possible to make this fixture call as a function by screenshot("process-list.png")?
If I want to take screenshots at different stages of the test, how can I do it with this fixture?

The whole purpose of this fixture is to be called at the end of the test, regardless of whether test succeeded or not. So if you need to take screenshot somewhere in the middle, you should still use selenium driver directly as before. Does that make sense?

unkcpz commented:
> So if you need to take screenshot somewhere in the middle, you should still use selenium driver directly as before. Does that make sense?

It makes sense. That also means with this fixture it is expected to have only one screenshot for a test case. I'd then suggest setting the filename of the screenshot based on the function name, and then the fixture called inside the test function `screenshot["name"] = "<test_function_name>.png"` can be removed. Is that possible to implement?

danielhollas commented:
That's a good idea and I actually did try to implement exactly that, but
couldn't make it work. I'll try a bit more.

Dne čt 5. 1. 2023 10:03 dop. uživatel Jusong Yu ***@***.***>
napsal:

> So if you need to take screenshot somewhere in the middle, you should
> still use selenium driver directly as before. Does that make sense?
>
> It makes sense. That also means with this fixture it is expected to have
> only one screenshot for a test case. I'd then suggest setting the filename
> of the screenshot based on the function name, and then the fixture called
> inside the test function screenshot["name"] = "<test_function_name>.png"
> can be removed. Is that possible to implement?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/aiidalab/aiidalab-widgets-base/pull/421#issuecomment-1372011832>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACIY64IGTK7GF3EJ27BSLZTWQ2L7LANCNFSM6AAAAAATAWLULU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

yakutovicha reviewed (commented):
Thanks, @danielhollas. I also prefer this approach w.r.t. #422.

I have just a few suggestions.

danielhollas commented:
@yakutovicha @unkcpz thanks for the reviews! I've modified the new fixture so that the screenshot name is determined automatically from the test function name. For example, this test

```pytest
def test_something_important(final_screenshot):
       pass
```

will automatically generate a screenshot titled `something_important.png`.

I've also changed the fixture name from `screenshot` to `final_screenshot` to make it clear that this is a teardown fixture.

danielhollas commented on tests/conftest.py:
```diff
@@ -91,14 +91,15 @@ def _selenium_driver(nb_path):
+    Screenshot name is generated from the test function name
+    by stripping the 'test_' prefix
+    """
+    screenshot_name = f"{request.function.__name__[5:]}.png"
```
I don't think lstrip here does what you think:

https://stackoverflow.com/questions/4148974/removing-a-prefix-from-a-string

unkcpz replied:
I see, then just ignore my request.

danielhollas replied:
Note that in Python 3.9 one could use the `removeprefix` string method for exactly this, but we'd need to drop support for Python 3.8, at least in tests. Probably not worth it now.

unkcpz reviewed (approved):
Looks good to me.
